File: Algorithm/46-permutations.py
# 46. Permutations
# Given a collection of distinct integers, return all possible permutations.

# Runtime: 44 ms, faster than 100% of submissions
# Memory usage: 6.6 MB, less than 91.98% of submissions

class Solution:                
    def permute(self, nums):
        
        def swap(nums, a, b):
            temp = nums[a]
            nums[a] = nums[b]
            nums[b] = temp
        
        result = [nums]
        for i in range(0, len(nums)):
            l = len(result)
            for j in range(i+1, len(nums)):
                for k in range(0, l):
                    dupe = result[k].copy()
                    swap(dupe, i, j)
                    result.append(dupe)
        
        
        return result
    
    
                

File: Algorithm/46-permutations.java
// 46. Permutations
// 2 ms, faster than 99.89% of Java submissions
// 27.5 MB, less than 15.56% of Java submissions

class Solution {
    public List<List<Integer>> permute(int[] nums) {
        List<List<Integer>> result = new ArrayList<>();
        List<Integer> init = new ArrayList<>();
        for (int n = 0; n < nums.length; n++) {
            init.add(nums[n]);
        }
        result.add(init);
        for (int i = 0; i < nums.length; i++) {
            int l = result.size();
            for (int j = i + 1; j < nums.length; j++) {
                for (int k = 0; k < l; k++) {
                    List<Integer> dupe = new ArrayList<>(result.get(k));
                    swap(dupe, i, j);
                    result.add(dupe);
                }
            }
        }
        return result;
        
    }
    public void swap(List<Integer> nums, int a, int b) {
        int temp = nums.get(a);
        nums.set(a, nums.get(b));
        nums.set(b, temp);
    }
}